File: Solutions/830-largest-triangle-area/largest-triangle-area.py
from typing import List  # For type hinting
import itertools         # To generate combinations of 3 points easily

class Solution:
    def largestTriangleArea(self, points: List[List[int]]) -> float:
        
        # Helper function to compute the area of a triangle using three points
        def triangle_area(p1, p2, p3):
            # Unpack each point into x and y coordinates
            x1, y1 = p1
            x2, y2 = p2
            x3, y3 = p3

            # Use the Shoelace formula to compute the area of the triangle
            # Formula: 0.5 * |x1*(y2 - y3) + x2*(y3 - y1) + x3*(y1 - y2)|
            return abs(x1 * (y2 - y3) + 
                       x2 * (y3 - y1) + 
                       x3 * (y1 - y2)) / 2

        # Initialize the maximum area found so far to 0
        max_area = 0.0

        # Generate all unique combinations of 3 different points
        for p1, p2, p3 in itertools.combinations(points, 3):
            # Calculate the area of the triangle formed by p1, p2, and p3
            area = triangle_area(p1, p2, p3)

            # Update max_area if this triangle's area is larger
            max_area = max(max_area, area)

        # Return the maximum area found
        return max_area

File: Solutions/830-largest-triangle-area/largest-triangle-area.ex
defmodule Solution do
  @spec largest_triangle_area(points :: [[integer]]) :: float
  def largest_triangle_area(points) do
    n = length(points)

    # Helper function to calculate the area of a triangle given three points
    area = fn [x1, y1], [x2, y2], [x3, y3] ->
      abs(x1*(y2 - y3) + x2*(y3 - y1) + x3*(y1 - y2)) / 2.0
    end

    # Initialize max_area to 0
    max_area = 0.0

    # Loop over all combinations of three different points
    max_area =
      for i <- 0..(n-3), j <- (i+1)..(n-2), k <- (j+1)..(n-1), reduce: 0.0 do
        acc ->
          a = Enum.at(points, i)
          b = Enum.at(points, j)
          c = Enum.at(points, k)

          # Calculate area of triangle formed by points a, b, c
          current_area = area.(a, b, c)

          # Keep the maximum
          max(acc, current_area)
      end

    max_area
  end
end
